### ISSUE TITLE
System.Reflection.TargetInvocationException after run finished

### ISSUE THREAD

**Issue by undron** [nunit/nunit#1958]
After running tests (`/packages/NUnit.ConsoleRunner.3.5.0/tools/nunit3-console.exe workspace/Tests/ED_DC_Tests.dll --where "cat == _Acceptance" --work workspace --inprocess `) there is error:

```
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at NUnit.Framework.Internal.PropertyBag.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.Test.PopulateTestNode(TNode thisNode, Boolean recursive)
   at NUnit.Framework.Internal.TestMethod.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Internal.TestResult.AddToXml(TNode parentNode, Boolean recursive)
   at NUnit.Framework.Api.FrameworkController.RunTests(ICallbackEventHandler handler, String filter)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeConstructorInfo.Invoke(BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.RuntimeType.CreateInstanceImpl(BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(Type type, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes)
   at System.Activator.CreateInstance(String assemblyString, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo, StackCrawlMark& stackMark)
   at System.Activator.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityInfo)
   at System.AppDomain.CreateInstance(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at System.AppDomain.CreateInstanceAndUnwrap(String assemblyName, String typeName, Boolean ignoreCase, BindingFlags bindingAttr, Binder binder, Object[] args, CultureInfo culture, Object[] activationAttributes, Evidence securityAttributes)
   at NUnit.Engine.Drivers.NUnit3FrameworkDriver.Run(ITestEventListener listener, String filter)
   at NUnit.Engine.Runners.DirectTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
   at NUnit.Engine.Runners.MasterTestRunner.RunTests(ITestEventListener listener, TestFilter filter)
   at NUnit.Engine.Runners.MasterTestRunner.Run(ITestEventListener listener, TestFilter filter)
   at NUnit.ConsoleRunner.ConsoleRunner.RunTests(TestPackage package, TestFilter filter)
   at NUnit.ConsoleRunner.Program.Main(String[] args)
```

The error is in some test, because other categories runs good. All tests runs in parallel. Any suggesting to find buggy case?

**Comment by undron** [comment]
I have found this case. The minimum working example is:
```c#
    public class Class1
    {
        [Test]
        public void test1()
        {
            TestContext.CurrentContext.Test.Properties.Add("CaseName", null);
        }
    }
```

**Comment by CharliePoole** [comment]
Actually we have two issues here:

1. The obvious one: a crash is caused by trying to add a property with a null value to the XML. This should probably be prevented by the PropertyBag class itself, since properties are intended to have non-null values. A secondary check could be included in the XML generation so that any null property does not get added to the XML.

2. The more subtle error is that it has never been intended for users to be able to add properties to a test at execution time or to modify the test properties in any way. This is possible because of an error in the TestAdapter class, which exposes the properties of the internal test instance, permitting possible destructive changes to the properties. I think this has to be fixed as well, making it impossible for users to add or change properties at runtime. 

Unfortunately, it seems as if @undron may be making use of this "feature" and other  could have discovered it as well. We need to give some thought as to how we can replace this facility. One approach would be a PropertyBagAdapter that allowed addition of new properties, as illustrated here, but not changes to existing ones. 

Another approach, which I've discussed elsewhere, would be the addition of properties to the test result, which would only be created at the time of executing the test. That would keep the dynamically set properties separate from the static properties that are associated with the test as a result of attributes on the test method. Additionally, the new set of properties could use a standard IDictionary interface rather than the idiosyncratic IPropertyBag interface we use internally.

-----

Let's use this issue to fix number 1 above and discuss options for number 2. We can create a new issue once we decide on a long-term approach.

**Comment by undron** [comment]
I know, it's my bad to try to add null value to properties, but adding it at runtime is very useful feature. We use it to add localized case name to TestResult.xml and then to generate JUnit report with this names by slightly modified https://github.com/nunit/nunit-transforms/tree/master/nunit3-junit

**Comment by jnm2** [comment]
@CharliePoole For 2: what other things besides properties can be mutated off `ITest` and `TestContext`? Eventually, would it make sense to expose a readonly version so that `Add` is not present at compile time? (Name is just for the sake of example but `IReadOnlyTest` could have `IReadOnlyDictionary `Properties` along with read-only replacements for any other mutable properties.)

Are there any real use cases where modifying properties is the best workaround at this time? They could be addressed as they come up through a more idiomatic API than setting magic strings in a property bag.

**Comment by CharliePoole** [comment]
TestContext is designed to be read-only and no ITest interface is ever returned. The problem here is that the Properties getter is returning a mutable object, but that's rather easy to fix. I'm fairly confident that this is the only one but at the same time I agree it makes sense to check all of them.
 
That's as far as I think this issue should go - support for adding properties is a different matter to be dealt with on a different time frame.

Regarding "magic strings" I don't really understand what you are saying. We're not talking properties in the C# sense here, but test properties, which can have any name that users care to give them. The whole point of a PropertyBag is that NUnit doesn't always know what the names of the properties are. There are, of course some built-in properties that NUnit supports internally, but the use of string keys there is a matter of decoupling the implementation of the engine from that of the framework.

**Comment by jnm2** [comment]
:+1:

> Regarding "magic strings" I don't really understand what you are saying.

Just that for example an eventual API to set localized case names should probably hide the fact that strings are used internally for decoupling.

**Comment by CharliePoole** [comment]
I triaged this issue as dealing with the "bug" portion only - item 1 of the 2 in my earlier comment, i.e.

There are probably two things to fix even for the bug...

1. PropertyBag should handle adding a null with a more graceful error.

2. The Properties exposed to the user should be an adapter, just as we expose a test adapter and a result adapter to the user rather than the real test or result. It should not allow adding or modifying anything.

This will harm some users, who are using what was an unintended "feature" for their own purposes. That's not all bad, since they will tell us what they need and we can set it up to work correctly. Right now, we know that some users need what I call "result" properties, some need to be able to set properties in the context and others are trying to localize names of tests. I'll check to see if we have issues for those things already and add them if we don't.

**Comment by rprouse** [comment]
@CharliePoole better solution for the second issue than my non-solution in #2278 😄

**Comment by CharliePoole** [comment]
Ha! I looked at the issues in the wrong order. :smiley:

**Comment by OmicronPersei** [comment]
Just checking in, is the intention still to only fix @CharliePoole 's number 1 issue or to fix both on this PR?

**Comment by jnm2** [comment]
My preference is to fix both, but already several choices come to mind that we'll ahve to make about what kinds of breaking changes we want to make. So can we do them separately and get the fix out soonest?

**Comment by OmicronPersei** [comment]
Well, number 2 looks simple enough to get it in with this as well pretty quickly.  Do we want our suggested `PropertyBagAdapter` to have all the same functionality as a regular `PropertyBag`, including adding elements to it, etc?  Or just serve as a glorified Dictionary that has accessors similar to `PropertyBag`?

**Comment by jnm2** [comment]
Here's an example of the kind of question I'd want the whole team to review:

Are we willing to change `ITest.Properties` to return an `IReadOnlyPropertyBag` instead of an `IPropertyBag`? That would cause this code to start failing:

```cs
ExtractSpecialProperty(TestContext.CurrentContext.Test.Properties);

// ...

Foo ExtractSpecialProperty(IPropertyBag properties)
{
    // ...
}
```

**Comment by OmicronPersei** [comment]
We could have `IPropertyBag` inherit from `IReadOnlyPropertyBag` where `IPropertyBag` supports setters in addition to the read only version?

**Comment by jnm2** [comment]
Sure, but either way, if you think of me being a long-time NUnit user with the code I showed in a nexample, when I upgrade to NUnit 3.10 I start gettinn compiler errors on the line `ExtractSpecialProperty(TestContext.CurrentContext.Test.Properties);`. If we break people's code, we need to be sure that's the best option and warn people in the release notes.

**Comment by CharliePoole** [comment]
First of all, as to what I meant in introducing two numbered issues is that they could and probably should be two PRs. They solve two different problems and have entirely different levels of urgency. Also, the second point will probably lead to a design discussion. IME PRs have very little fixed overhead so we shouldn't hesitate to do two of them when needed.

With regard to the second point, I would tend to expose a read-only property bag of our own construction. It would simply not have the setters present that are on IPropertyBag. Since we only use it in one place, it can easily be a one-off class rather than an interface. But as I said, I'd do this in a follow-up PR.
